### PR TITLE
[feature] grants for additional resources

### PR DESCRIPTION
**Pull request by a129722**
This PR adds functionality for grants of type:

- ADD SEARCH OPTIMIZATION on schema
- future grants on stages
- external tables
- materialized views
- streams
- procedures
- functions
- file formats
- sequences

## Test Plan
* [x] acceptance tests for future grants, normal grants require resource implementations
* [x] unit tests for all new grant types

### CODE REVIEW

File: pkg/resources/external_table_grant.go
Comment (ryanking) [inline pkg/resources/external_table_grant.go:115]: I think we would have a better experience if we pull these up to the schema level with [ValidateDiagFunc](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Schema.ValidateDiagFunc)

File: pkg/resources/external_table_grant_acceptance_test.go
Comment (ryanking) [inline pkg/resources/external_table_grant_acceptance_test.go:14]:
```suggestion
func TestAccExternalTableGrant_future(t *testing.T) {
```

File: pkg/resources/file_format_acceptance_test.go
Comment (ryanking) [inline pkg/resources/file_format_acceptance_test.go:14]:
```suggestion
func TestAccFileFormatGrant_future(t *testing.T) {
```

File: pkg/resources/file_format_acceptance_test.go
Comment (ryanking) [inline pkg/resources/file_format_acceptance_test.go:17]:
```suggestion
	databaseName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
	schemaName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
	roleName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
```

File: pkg/resources/file_format_grant.go
Comment (ryanking) [inline pkg/resources/file_format_grant.go:20]:
```suggestion
		Description: "The name of the file format on which to grant privileges immediately (only valid if on_future is false).",
```

File: pkg/resources/file_format_grant.go
Comment (ryanking) [inline pkg/resources/file_format_grant.go:108]: As with above, it would be great to pull these up to the schema level.

**Review by ryanking — CHANGES_REQUESTED** [review]
Thanks for this, majorly helpful work.

I only read through the first dozen or so files, but I think most of my comments there apply to the rest. Take a look and click the re-request review when ready and I will review the rest.

File: pkg/resources/database_acceptance_test.go
Comment (a129722) [reply pkg/resources/database_acceptance_test.go:18]: This change is to make the acceptance tests compatible with accounts that have [QUOTED_IDENTIFIERS_IGNORE_CASE](https://docs.snowflake.com/en/sql-reference/parameters.html#quoted-identifiers-ignore-case) set to `true`.  This will cause the any identifiers used to be represented in all upper-case within any system catalog-based queries.

File: pkg/resources/external_table_grant.go
Comment (a129722) [reply pkg/resources/external_table_grant.go]: Updated

File: pkg/resources/external_table_grant.go
Comment (a129722) [reply pkg/resources/external_table_grant.go:64]: To answer the two ways to interpret the question... 

1. Yes, it only really conflicting when `external_table_name` is set to true, but there's no good way aside from this to enforce within the schema itself.  
2. Future grants are not supported with shares

File: pkg/resources/external_table_grant.go
Comment (a129722) [reply pkg/resources/external_table_grant.go]: Updated

File: pkg/resources/external_table_grant.go
Comment (a129722) [reply pkg/resources/external_table_grant.go]: Updated

File: pkg/resources/external_table_grant_acceptance_test.go
Comment (a129722) [reply pkg/resources/external_table_grant_acceptance_test.go]: So I was looking at other examples of format for test names and the suggestion above conflicts with others.  I ended up with the name `TestAcc_ExternalTableFutureGrant` and applied that elsewhere.  Let me know if you'd like something different.

File: pkg/resources/file_format_acceptance_test.go
Comment (a129722) [reply pkg/resources/file_format_acceptance_test.go]: See above

File: pkg/resources/file_format_acceptance_test.go
Comment (a129722) [reply pkg/resources/file_format_acceptance_test.go:17]: See above

File: pkg/resources/file_format_grant.go
Comment (a129722) [reply pkg/resources/file_format_grant.go]: Updated

File: pkg/resources/file_format_grant.go
Comment (a129722) [reply pkg/resources/file_format_grant.go]: Done

**Comment by a129722** [comment]
@ryanking - this is ready for re-review.  i've also integrated changes from `main` and added examples for all resources.

**Review by ryanking — CHANGES_REQUESTED** [review]
This is another partial review. I am going to hold off on reviewing more until we land https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/333, which will requires some rebasing here.

File: pkg/resources/external_table_grant.go
Comment (a129722) [reply pkg/resources/external_table_grant.go:21]: A validation function only operates on the value itself, so I think the validation would need to go within the Create function like we had before.

Comment (ryanking) [reply pkg/resources/external_table_grant.go:21]: Yeah you are right. I will complain to the team at hashicorp about the lack of a good validation hook for this.

File: pkg/resources/database_acceptance_test.go
Comment (a129722) [reply pkg/resources/database_acceptance_test.go:18]: That is correct

Comment (ryanking) [reply pkg/resources/database_acceptance_test.go:18]: Ok, if this is blocking we can do it for now and I will think of more sustainable ways to do this.

**Review by ryanking — CHANGES_REQUESTED** [review]
Can you merge the main branch and resolve conflicts?

Also, note that the refactor in https://github.com/chanzuckerberg/terraform-provider-snowflake/commit/ffe977ad2bcd3165b1821ae79f44dec7246aeb7f is going to require some changes to this PR to match.

**Comment by a129722** [comment]
@ryanking - main was merged again and I updated the code to remove the `ConflictsWith` schema declarations and added in the checks within the `Create` methods.

**Comment by a129722** [comment]
@ryanking  - I was wondering if we could push to get this reviewed/integrated soon.  There's been a good amount change on the main branch and re-integrating everything has been time consuming on my end.

**Comment by ryanking** [comment]
@a129722 yeah, as soon as you can resolve the merge conflicts we will take a look

**Comment by a129722** [comment]
@ryanking  - i merged in main.  not sure why the check-mod step is failing.  When I build/test locally, there are no changes that show up.

**Comment by ryanking** [comment]
Yeah there is something weird going on with `check-mod`. If you run `go get -u golang.org/x/tools` and `go mod tidy` it should fix it (not sure why we recently had to start doing this).

**Comment by a129722** [comment]
thanks for the tip - looks like things are passing now

File: pkg/resources/file_format_acceptance_test.go
Comment (ryanking) [inline pkg/resources/file_format_acceptance_test.go:1]: Should be name `file_format_grant_acceptance_test.go`

File: pkg/resources/file_format_acceptance_test.go
Comment (ryanking) [inline pkg/resources/file_format_acceptance_test.go:14]:
```suggestion
func TestAcc_FileFormatGrantFutureGrant(t *testing.T) {
```

**Review by ryanking — CHANGES_REQUESTED** [review]
Very minor changes and then this is good to merge.

**Comment by a129722** [comment]
Done!

**Review by ryanking — APPROVED** [review]

**Comment by a129722** [comment]
Thanks, @ryanking !

**Pull request merged**
